### PR TITLE
WIP: 0 GNN layers on last IPU + Fingerprinting

### PR DESCRIPTION
**Pull request by callumm-graphcore**
## Changelogs

- Allow having 0 GNN layers (i.e. only the task heads) on the last IPU in a pipeline split

---

_Checklist:_

- [ ] _Was this PR discussed in an issue? It is recommended to first discuss a new feature into a GitHub issue before opening a PR._
- [ ] _Add tests to cover the fixed bug(s) or the new introduced feature(s) (if appropriate)._
- [ ] _Update the API documentation is a new function is added, or an existing one is deleted._
- [ ] _Write concise and explanatory changelogs above._
- [ ] _If possible, assign one of the following labels to the PR: `feature`, `fix` or `test` (or ask a maintainer to do it for you)._

---

_discussion related to that PR_

### CODE REVIEW

File: graphium/nn/architectures/global_architectures.py [hunk 1]
```diff
@@ -1341,9 +1352,28 @@ def forward(self, g: Batch) -> Tensor:
...
+                return g, extras_to_return
+
+            final_output, task_head_extras_to_return = self.task_heads.forward(
+                g, extra_return_names=extra_return_names
+            )
+            extras_to_return.update(task_head_extras_to_return)
+
+            return final_output, extras_to_return
+
+        # Keep original code if no extra_return_names
```
Comment (s-maddrellmander) [inline graphium/nn/architectures/global_architectures.py:1376]: Maybe I'm missing something obvious, in which case apologies. 
Aren't you missing the original forward pass in the case the `extra_return_names` is missing?
```suggestion
        # Keep original code if no extra_return_names
        else:
             g = self.gnn.forward(g)
```

**Review by s-maddrellmander — APPROVED** [review]
One tiny question, apologies if I'm missing anything here. 
But otherwise the logic looks good to me. Interested to see the results of this. 

What exactly is the motivation with having no GNN layers on the final IPU, making more space for heads / fine-tuning ?

**Comment by callumm-graphcore** [comment]
Sorry Sam, this was more about me sharing some extra code with Kerstin than making a change to the library, I will look at your comments tomorrow

**Comment by s-maddrellmander** [comment]
No worries! Just saw there was an outstanding PR and glanced thought, ping me when there’s something concrete to review. :)

**Review by DomInvivo — APPROVED** [review]

**Comment by DomInvivo** [comment]
@callumm-graphcore, @s-maddrellmander , any news on this PR? It was never merged, but I suppose you want to merge it? Since March 12th.

**Comment by callumm-graphcore** [comment]
Hi @DomInvivo, we're no longer working on this so I've just closed both PRs

**Pull request closed without merging**